Consensus: Turn CBlockIndex::GetMedianTimePast into independent function

jtimon opened the pull request:
ContextualCheckBlockHeader() [which is needed to verify a block header] depends on CBlockIndex::GetMedianTimePast but the goal is for verify block header is to be independent from chain.o and CBlockIndex C++ class [which cannot be exposed in the consensus C API].
This is also contained in #5995.

jtimon committed:
Consensus: Refactor: Turn CBlockIndex::GetMedianTimePast into independent function

...and CBlockIndex::nMedianTimeSpan into Consensus::Params::nMedianTimeSpan

jtimon commented:
Needed rebase.

jtimon commented:
Needed rebase.

sipa commented:
utACK

sipa commented:
Needs rebase.

jtimon commented:
@sipa I know, but I'm waiting for #6051 to be merged since it will require another rebase after that.

jtimon commented:
Rebased

petertodd commented on src/consensus/blockverify.cpp:
```diff
@@ -0,0 +1,25 @@
+
+#include "chain.h"
+
+#include <algorithm>  
```
Nit: tailing whitespace

petertodd commented:
Modulo nit, (somewhat tested) ACK.

jtimon commented:
@petertodd 's nit fixed.

btcdrak commented:
utACK

jtimon commented:
Closing for now.

theuni committed:
consensus: don't define MAX_STANDARD_TX_SIGOPS in terms of block size

theuni committed:
consensus: teach CheckTransaction to check for an arbitrary max tx size

This is a no-op change. For now, everything passes MAX_BLOCK_SIZE, so the
result matches what it would've before.

Tests use a value of std::numeric_limits<uint64_t>::max() where necessary, to ensure that they're never
rejected when size isn't being tested.

theuni committed:
consensus: Move consensus constants into Consensus::Params and consensus.h (as functions)

The following are now tied to a chain rather than being defined as global
constants. Their values have not changed.

nMinTxSize
nMaxBlockSize
nMaxTxSize
nMaxBlockSigops
nCoinbaseMaturity

Also, for free (diff-wise):

Blocksize: Turn MAX_BLOCK_SIZE (nMaxBlockSize) and MAX_BLOCK_SIGOPS (nMaxBlockSigops) into functions

...which take Consensus::Params as parameter
This will be convenient to reduce the diff of any proposal that changes the blocksize as a hardfork

theuni committed:
consensus: teach ExtractMatches to check for an arbitrary max transaction number

This is a no-op change. For now, everything passes MAX_BLOCK_SIZE / 60, so the
result matches what it would've before.

Tests use a number equal to the number of transactions where necessary,
to ensure that they're never rejected when blocksizesize isn't being tested.

jtimon commented:
Rebased and reopened.

jtimon committed:
Consensus: MOVEONLY: Most important consensus function declarations to consensus/consensus.h

jtimon committed:
Consensus: Chainparams: Explicit Consensus::Params for consensus functions:

-CheckBlockHeader
-ContextualCheckBlockHeader
-CheckBlock
-ContextualCheckBlock

Also add nTime parameter to CheckBlockHeader and CheckBlock.
Also use the oportunity to rename the functions inside the Consensus namespace.

jtimon commented:
Rebased on top of #6591. Still had merging conflicts so rebased on top of master after that.

dcousens commented:
utACK for c85ded0 after nit is addressed

jtimon commented:
Closing in favor of #6672 .
